File: Java/src/Permutations.java
import java.util.ArrayList;
import java.util.Arrays;

public class Solution
{
  public List<List<Integer>> permute(int[] num)
  {
    int length = num.length;
    List<List<Integer>> ret = new ArrayList<>();
    if (length == 0)
      return ret;
    Arrays.sort(num);
    List<Integer> numList = new ArrayList<>();
    numList.add(num[0]);
    ret.add(numList);

    for (int i = 1; i < length; i++)
    {
      List<List<Integer>> newList = new ArrayList<>();
      for (int j = 0; j <= i; j++)
      {
        for (List<Integer> k : ret)
        {
          List<Integer> copyOfK = new ArrayList<>(k);
          copyOfK.add(j, num[i]);
          newList.add(copyOfK);
        }
      }
      ret = newList;
    }
    return ret;
  }
}

File: Python/Permutations.py

class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    def permute(self, num):
        length = len(num)
        ret = []
        num.sort()
        reversenum = num[::-1]
        innerdata = num[:]
        ret.append(innerdata[:])
        i = 0
        while innerdata != reversenum:
            for i in range(length-1, 0, -1):
                if innerdata[i-1] < innerdata[i]:
                    newlist = innerdata[:i-1:-1]
                    
                    for l in range(len(newlist)):
                        if innerdata[i-1] < newlist[l]:
                            temp = innerdata[i-1]
                            innerdata[i-1] = newlist[l] 
                            newlist[l] = temp
                            innerdata = innerdata[:i] + newlist
                            break
                    break
            ret.append(innerdata[:]) 
        return ret 
        